File: test/workspaceAsk.test.mjs
/**
 * The 0.15.0 ask in the TS twin (openspec: refocus-chat-attachments §1.8): a
 * question answered from a conversation's ATTACHMENTS, with no vault anywhere
 * in the path.
 *
 * PARITY: the twin of native/crates/lighthouse-core/tests/workspace_ask_test.rs
 * — the same claims, over the same corpus swap. Model-free by construction (a
 * config with no provider answers extractively), so this runs zero-network.
 *
 * Run: `node --test test/workspaceAsk.test.mjs`
 */
import { test } from "node:test";
import assert from "node:assert/strict";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { register } from "node:module";

register("./_ts-extensionless-hook.mjs", import.meta.url);

const ws = await import("../src/server/workspace.ts");
const synth = await import("../src/server/synth.ts");

/** Fresh app-state root per test — appStateDir() re-reads the env var. */
function freshState(tag) {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), `lh-ask-${tag}-`));
  process.env.LIGHTHOUSE_APP_STATE_DIR = dir;
  // A vault dir that exists but is EMPTY: if any leg of the pipeline reached
  // for the vault instead of the workspace, it would find nothing and the
  // citation assertions below would fail rather than silently pass.
  process.env.VAULT_DIR = fs.mkdtempSync(path.join(os.tmpdir(), `lh-ask-${tag}-vault-`));
  return dir;
}

/** Drain the pipeline over one conversation's corpus; return text + citations. */
async function ask(question, conversationId) {
  const chunks = [];
  let text = "";
  for await (const c of synth.answerPipeline(
    question,
    [],
    [],
    [],
    { providerId: undefined, modelId: undefined, apiKey: undefined },
    {},
    [],
    new synth.Corpus(conversationId),
  )) {
    text += c.delta ?? "";
    chunks.push(c);
  }
  const cited = [
    ...new Set(chunks.flatMap((c) => (c.references ?? []).map((r) => r.name))),
  ].sort();
  return { text, cited };
}

test("an ask answers from its conversation's attachments alone", async () => {
  freshState("scope");
  ws.attach(
    "conv-1",
    "quarterly.md",
    Buffer.from("# Q3 revenue\nNortheast revenue rose sharply this quarter, led by hardware.\n"),
  );
  // A file attached to a DIFFERENT conversation must stay invisible here.
  ws.attach("conv-other", "secret.md", Buffer.from("# Secret\nProject Kestrel ships in May.\n"));

  const { cited } = await ask("What happened to Northeast revenue?", "conv-1");
  assert.ok(cited.includes("quarterly.md"), `the attachment is cited: ${cited}`);
  assert.ok(
    !cited.includes("secret.md"),
    "another conversation's attachment never joins the corpus",
  );

  // A conversation with nothing attached has nothing to cite.
  const empty = await ask("What happened to Northeast revenue?", "conv-empty");
  assert.deepEqual(empty.cited, [], "no attachments ⇒ no sources");
});

test("the same question over the vault corpus does not see the attachments", async () => {
  freshState("vault");
  ws.attach("conv-1", "quarterly.md", Buffer.from("# Q3\nNortheast revenue rose sharply.\n"));
  // A null conversation id selects the LEGACY vault corpus, which is empty
  // here — proof the two arms are actually distinct and the workspace answer
  // above came from the workspace.
  const { cited } = await ask("What happened to Northeast revenue?", null);
  assert.deepEqual(cited, [], "the vault arm sees an empty vault, not the workspace");
});

test("retrieval ranks within the conversation: the relevant attachment wins", async () => {
  freshState("rank");
  ws.attach("conv-1", "revenue.md", Buffer.from("# Revenue\nNortheast revenue rose sharply.\n"));
  ws.attach("conv-1", "hiring.md", Buffer.from("# Hiring\nWe opened twelve roles in support.\n"));

  const { cited } = await ask("What happened to Northeast revenue?", "conv-1");
  assert.ok(cited.includes("revenue.md"), `the matching file is cited: ${cited}`);
});

File: native/crates/lighthouse-core/tests/workspace_ask_test.rs
//! The 0.15.0 ask (openspec: refocus-chat-attachments): a question answered
//! from a conversation's ATTACHMENTS, with no vault anywhere in the path.
//!
//! This is the end-to-end proof that the corpus swap works: the same
//! `answer_pipeline` that serves the vault serves a workspace corpus, the
//! answer cites the attached file, an unattached file is invisible, and the
//! answer cache replays across conversations that hold byte-identical bytes.
//! Model-free by construction — a device provider with no key answers
//! extractively, so this runs with zero network.

mod common;

use futures::StreamExt;
use lighthouse_core::contracts::ChatChunk;
use lighthouse_core::llm::ModelCfg;
use lighthouse_core::synth::{answer_pipeline, Corpus};
use lighthouse_core::workspace;

/// A device config with no key: the pipeline answers extractively.
fn device_cfg() -> ModelCfg {
    ModelCfg { provider_id: Some("local".into()), ..ModelCfg::default() }
}

async fn ask(question: &str, conversation: &str) -> (String, Vec<ChatChunk>) {
    let mut stream = answer_pipeline(
        question.to_string(),
        vec![],
        vec![],
        vec![],
        device_cfg(),
        Default::default(),
        Default::default(),
        vec![],
        Corpus { conversation_id: Some(conversation.to_string()) },
    );
    let (mut text, mut chunks) = (String::new(), Vec::new());
    while let Some(c) = stream.next().await {
        text.push_str(&c.delta);
        chunks.push(c);
    }
    (text, chunks)
}

fn cited(chunks: &[ChatChunk]) -> Vec<String> {
    let mut names: Vec<String> = chunks
        .iter()
        .filter_map(|c| c.references.as_ref())
        .flatten()
        .map(|r| r.name.clone())
        .collect();
    names.sort();
    names.dedup();
    names
}

#[tokio::test]
async fn an_ask_answers_from_its_conversation_attachments_alone() {
    let dir = tempfile::tempdir().unwrap();
    let _guard = common::lock_env(dir.path());

    workspace::attach(
        "conv-1",
        "quarterly.md",
        b"# Q3 revenue\nNortheast revenue rose sharply this quarter, led by hardware.\n",
    )
    .unwrap();
    // A file attached to a DIFFERENT conversation must stay invisible here.
    workspace::attach("conv-other", "secret.md", b"# Secret\nProject Kestrel ships in May.\n")
        .unwrap();

    let (text, chunks) = ask("What happened to Northeast revenue?", "conv-1").await;
    assert!(
        cited(&chunks).contains(&"quarterly.md".to_string()),
        "the attachment is cited: {:?}",
        cited(&chunks)
    );
    assert!(
        !cited(&chunks).contains(&"secret.md".to_string()),
        "another conversation's attachment never joins the corpus"
    );
    assert!(!text.trim().is_empty(), "an extractive answer still says something");

    // A conversation with nothing attached has nothing to cite.
    let (_, empty) = ask("What happened to Northeast revenue?", "conv-empty").await;
    assert!(cited(&empty).is_empty(), "no attachments ⇒ no sources: {:?}", cited(&empty));
}

#[tokio::test]
async fn identical_attachments_replay_across_conversations() {
    let dir = tempfile::tempdir().unwrap();
    let _guard = common::lock_env(dir.path());
    let bytes = b"# Q3 revenue\nNortheast revenue rose sharply this quarter.\n";
    workspace::attach("conv-a", "quarterly.md", bytes).unwrap();
    workspace::attach("conv-b", "quarterly.md", bytes).unwrap();

    // The key is over content hashes, so two conversations holding identical
    // bytes key identically — the portability the vault-era global digest
    // could never offer.
    let q = "What happened to Northeast revenue?";
    let key_a =
        lighthouse_core::answer_cache::workspace_cache_key(Some("conv-a"), q, Some("local"), None, &[]);
    let key_b =
        lighthouse_core::answer_cache::workspace_cache_key(Some("conv-b"), q, Some("local"), None, &[]);
    assert_eq!(key_a, key_b, "identical bytes ⇒ identical key");

    // Store an answer under conv-a's key, then ask in conv-b: the pipeline
    // replays it verbatim and stamps it as cached. (A real answer is stored
    // here directly because a model-free container answers "model unavailable",
    // which the engine deliberately refuses to cache.)
    lighthouse_core::answer_cache::reset_store();
    lighthouse_core::answer_cache::insert(
        &key_a,
        lighthouse_core::answer_cache::CachedAnswer {
            key: String::new(),
            created_ms: lighthouse_core::config::now_ms(),
            text: "Northeast revenue rose sharply, led by hardware.".to_string(),
            references: vec![],
            analytics: None,
            meta: lighthouse_core::contracts::ChunkMeta {
                origin: "device".to_string(),
                excerpt_count: 1,
                source_file_count: 1,
                cached_at: None,
                cost: None,
                manifest: None,
                chart: None,
                table: None,
            },
        },
        Default::default(),
    );

    let (text, chunks) = ask(q, "conv-b").await;
    assert_eq!(
        text.trim(),
        "Northeast revenue rose sharply, led by hardware.",
        "the other conversation's stored answer replays verbatim"
    );
    assert!(
        chunks.iter().any(|c| c.meta.as_ref().and_then(|m| m.cached_at).is_some()),
        "and is stamped as a replay, not passed off as fresh"
    );

    // One changed byte in the attachments is a different key — no false replay.
    workspace::attach("conv-c", "quarterly.md", b"# Q3 revenue\nNortheast revenue fell.\n")
        .unwrap();
    let (fresh, _) = ask(q, "conv-c").await;
    assert_ne!(
        fresh.trim(),
        "Northeast revenue rose sharply, led by hardware.",
        "different bytes never replay another conversation's answer"
    );
}
